File: cses/leetcode/go/502.go
package main

import (
	"container/heap"
	"fmt"
	"sort"
)

type Project struct {
	Profit  int
	Capital int
}
type ByCapital []Project

func (a ByCapital) Len() int           { return len(a) }
func (a ByCapital) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByCapital) Less(i, j int) bool { return a[i].Capital < a[j].Capital }

type MaxHeap []int

func (m MaxHeap) Len() int           { return len(m) }
func (m MaxHeap) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
func (m MaxHeap) Less(i, j int) bool { return m[i] > m[j] }
func (m *MaxHeap) Push(x interface{}) {
	*m = append(*m, x.(int))
}
func (m *MaxHeap) Pop() interface{} {
	old := *m
	n := len(old)
	x := old[n-1]
	*m = old[0 : n-1]
	return x
}
func findMaximizedCapital(k int, w int, profits []int, capital []int) int {
	n := len(profits)
	projects := []Project{}
	for i := 0; i < n; i++ {
		projects = append(projects, Project{Profit: profits[i], Capital: capital[i]})
	}
	fmt.Println(projects)
	sort.Sort(ByCapital(projects))
	fmt.Println("Sorted projects", projects)
	maxHeap := &MaxHeap{} //this a pointer
	heap.Init(maxHeap)
	index := 0
	for i := 0; i < k; i++ {
		for index < n && projects[index].Capital <= w {
			heap.Push(maxHeap, projects[index].Profit)
			index++
		}
		if maxHeap.Len() == 0 {
			break
		}
		w += heap.Pop(maxHeap).(int)
	}
	//1. Create a max-heap and store all the projects and have a capital requirement atmost the current capital which I have
	//2.max-heap to be sorted on the basis of their profit

	return w
}

File: leetcode/cpp/502.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

struct Project
{
    int profit;
    int capital;
};

class Solution
{
public:
    int findMaximizedCapital(int k, int w, vector<int> &profits, vector<int> &capital)
    {
        int n = profits.size();
        vector<Project> projects;

        // Creating a vector of Project structs
        for (int i = 0; i < n; i++)
        {
            projects.push_back({profits[i], capital[i]});
        }

        // Sorting projects by their capital requirements
        sort(projects.begin(), projects.end(), [](const Project &a, const Project &b)
             { return a.capital < b.capital; });

        // Using a max heap to keep track of the most profitable projects that can be taken
        priority_queue<int> maxHeap;
        int index = 0;

        // Selecting at most k projects
        for (int i = 0; i < k; i++)
        {
            // Adding all projects that can be taken with the current capital to the max heap
            while (index < n && projects[index].capital <= w)
            {
                maxHeap.push(projects[index].profit);
                index++;
            }

            // If there are no projects that can be taken, break out of the loop
            if (maxHeap.empty())
                break;

            // Take the most profitable project
            w += maxHeap.top();
            maxHeap.pop();
        }

        return w;
    }
};
